zeppelin-server/src/test/java/org/apache/zeppelin/rest/ZeppelinSparkClusterTest.java

```diff
@@ -16,588 +16,630 @@
 import org.apache.zeppelin.interpreter.InterpreterResult;
 import org.apache.zeppelin.interpreter.InterpreterSetting;
+import org.apache.zeppelin.interpreter.SparkDownloadUtils;
 import org.apache.zeppelin.notebook.Note;
 import org.apache.zeppelin.notebook.Paragraph;
 import org.apache.zeppelin.scheduler.Job.Status;
 import org.apache.zeppelin.server.ZeppelinServer;
 import org.apache.zeppelin.user.AuthenticationInfo;
 import org.junit.AfterClass;
-import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
 
 /**
  * Test against spark cluster.
- * Spark cluster is started by CI server using testing/startSparkCluster.sh
  */
+@RunWith(value = Parameterized.class)
 public class ZeppelinSparkClusterTest extends AbstractTestRestApi {
-    AuthenticationInfo anonymous;
-
-    @BeforeClass
-    public static void init() throws Exception {
-        AbstractTestRestApi.startUp(ZeppelinSparkClusterTest.class.getSimpleName());
-    }
-
-    @AfterClass
-    public static void destroy() throws Exception {
-        AbstractTestRestApi.shutDown();
-    }
-
-    @Before
-    public void setUp() {
-        anonymous = new AuthenticationInfo("anonymous");
+  private static Logger LOGGER = LoggerFactory.getLogger(ZeppelinSparkClusterTest.class);
+
+  private String sparkVersion;
+  private AuthenticationInfo anonymous = new AuthenticationInfo("anonymous");
+
+  public ZeppelinSparkClusterTest(String sparkVersion) throws Exception {
+    this.sparkVersion = sparkVersion;
+    LOGGER.info("Testing SparkVersion: " + sparkVersion);
+    String sparkHome = SparkDownloadUtils.downloadSpark(sparkVersion);
+    setupSparkInterpreter(sparkHome);
+    verifySparkVersionNumber();
+  }
+
+  @Parameterized.Parameters
+  public static List<Object[]> data() {
+    return Arrays.asList(new Object[][]{
+        {"2.2.1"},
+        {"2.1.2"},
+        {"2.0.2"},
+        {"1.6.3"}
+    });
+  }
+
+  public void setupSparkInterpreter(String sparkHome) throws InterpreterException {
+    InterpreterSetting sparkIntpSetting = ZeppelinServer.notebook.getInterpreterSettingManager()
+        .getInterpreterSettingByName("spark");
+
+    Map<String, InterpreterProperty> sparkProperties =
+        (Map<String, InterpreterProperty>) sparkIntpSetting.getProperties();
+    LOG.info("SPARK HOME detected " + sparkHome);
+    if (System.getenv("SPARK_MASTER") != null) {
+      sparkProperties.put("master",
```

Comment:
should this be `spark.master `?

Reply:
Zeppelin use both spark.master & master which are the same.